### ISSUE TITLE
Password policy enforcement on front-end

### ISSUE THREAD

**Issue by adelefelicia** [indykite/jarvis-sdk-web#128]
**Created by: Stein Myrseth Sep 06 2021 at 14:51 CET**
Updated: Jul 08 2022 at 23:24 CET
ClickUp ID: IK-1445

The current UI-SDK is using its own UI to enable password entry on both account registration and forgotten password. The current implementation of password policy enforcements is set to 4 characters as minimum which is very weak.

Indyriot added their own frontend password validation component to the change password flow on the account management within the builtoriot platform. The current implementation can be checkout here or on any other tenants https://buildtoriot-dev.web.app/user/profile/account

The component we used as a base for our current deployment is this:
 The different parts of the policy should in the future be able to be requested from kite and feed to the component for frontend validation.

We are happy to share our current frontend code to be reused by the UI-SDK if possible.

The main story is to replace the current text input fields and the backend validation provided (4 characters) with the UI-SDK with our own look & feel component similar to the one we already have for "change password".

**Attachment(s):** 
https://github.com/indykite/clickup-attachments/blob/master/42e2c73f-76fd-424e-b390-6ad93cfa1a91.png
:link: Linked tasks: 
https://github.com/indykite/jarvis-sdk-node/issues/106 ,
https://github.com/indykite/jarvis-proto/issues/1065 ,

**Comment by adelefelicia** [comment]
**Commented by: @bayegan Sep 08 2021 at 20:25 CET**

@ Stein Myrseth : we will not be able to prioritize back-end development before october 25 2021. we'll revisit this task after that. we have APIs that can be used by RIOT to enforce password policy independent of UI-SDK. i'll add a pointer here. 



cc @dusan-kovacik-profiq

**Comment by adelefelicia** [comment]
**Commented by: @bayegan Sep 16 2021 at 23:46 CET**

@ Mateusz Grzywa @ Stein Myrseth : added to our next sprint, which ends October 12 (we'll release earlier if possible).

**Comment by adelefelicia** [comment]
**Commented by: @Petr Vecera Sep 17 2021 at 18:00 CET**

Currently they have validation like this. It will be not so easy to implement something like this in the UI SDK so it works properly with handling all the different states. I assume the best solution (for the developer) would be if we provided some customization object. This object could be like this:









```
{
 minLength: 8,
 maxLength: 120,
 lowerCase: 1, // means how many lower case characters - at least
 upperCase: 1, // means how many upper case characters - at least
 numbers: 4, // means how many numbers - at least 
 symbols: 2, // means how many symbols - at least 
}
```



With config like this we could handle the password validation on our own. And we could properly react and display proper messages and we would be also able to provide localization and other things. 



I assume that we could also offer multiple ways of validation the password. One could be the standard definition as mentioned. Another one could be usage of https://github.com/dropbox/zxcvbnhttps://github.com/dropbox/zxcvbn which applies modern principles to password strength. 

**Attachment(s):** 
https://github.com/indykite/clickup-attachments/blob/master/42e2c73f-76fd-424e-b390-6ad93cfa1a91.png

**Comment by adelefelicia** [comment]
**Commented by: @Petr Vecera Sep 17 2021 at 18:23 CET**

Currently we have released a version of SDK 1.2.0-beta-9 https://github.com/indykite/indykite-ui-sdk/pull/49/fileshttps://github.com/indykite/indykite-ui-sdk/pull/49/files and https://github.com/indykite/indykite-ui-sdk/pull/50https://github.com/indykite/indykite-ui-sdk/pull/50



Which allows the user to put validation function which stops the execution and no request is sent to indykite BE. However it doesn't show any error. IndyRiot wants to handle errors on their own end. 

This quite goes against the philosophy of the UI SDK. We offer 2 ways to use the SDK

- Build in UI - in this mode, everything should be handled by the UI SDK (including errors)
- Custom UI - in this mode the developer can build their own UI in a look and feel they want, they should also handle errors on their own 



Currently indyriot is using build in ui but mixes it with their own error handling. Which is not really the desired scenario.
